builtin/logical/postgresql/path_role_create.go

```diff
@@ -77,40 +81,52 @@ func (b *backend) pathRoleCreateRead(
 		Add(lease.Lease).
 		Format("2006-01-02 15:04:05-0700")
 
-	// Get our connection
+	// Get our handle
+	b.logger.Println("[TRACE] postgres/pathRoleCreateRead: getting database handle")
 	db, err := b.DB(req.Storage)
 	if err != nil {
 		return nil, err
 	}
 
 	// Start a transaction
+	b.logger.Println("[TRACE] postgres/pathRoleCreateRead: starting transaction")
 	tx, err := db.Begin()
 	if err != nil {
 		return nil, err
 	}
-	defer tx.Rollback()
+	defer func() {
+		b.logger.Println("[TRACE] postgres/pathRoleCreateRead: rolling back transaction")
+		tx.Rollback()
```

Comment:
For readonly transactions, `ROLLBACK` is the cheapest way to discard the transaction.  There are two ways to do this, actually.

The first would be to modify this slightly, something like:

```go
var commit bool
defer func() {
  if commit {
    err = tx.Commit()
    return
  }
  tx.Rollback()
}
```

And then change the `tx.Commit` on line `123` to be `commit=true`.  But that doesn't work well with the current function's flow and is complex with the interleaved `defer stmt.Close()`

A better approach would be to maintain local state for the existing database connection and handle things correctly in the `defer`'ed cleanup handler.  Something like:

```go
var txnOpen bool
var dirtyTxn bool
var stmtOpen bool
defer func() {
  if stmtOpen {
    stmt.Close()
  }
  if txnOpen {
    if dirtyTxn {
      err = tx.Commit()
      return
    }
    tx.Rollback()
  }
}
```

And then as various SQL is performed that writes, set `txnOpen=true; dirtyTxn=true`, etc.

CC @LLBennett